addons/binding/org.openhab.binding.xiaomivacuum/src/main/java/org/openhab/binding/xiaomivacuum/discovery/VacuumDiscovery.java

```diff
@@ -0,0 +1,196 @@
+ * and their token
+ *
+ * @author Marcel Verpaalen - Initial contribution
+ *
+ */
+public class VacuumDiscovery extends AbstractDiscoveryService implements ExtendedDiscoveryService {
+
+    /** The refresh interval for background discovery */
+    private static final long SEARCH_INTERVAL = 600;
+    private static final int TIMEOUT = 10000;
+    private ScheduledFuture<?> roboDiscoveryJob;
+    private DiscoveryServiceCallback discoveryServiceCallback;
+
+    private final Logger logger = LoggerFactory.getLogger(VacuumDiscovery.class);
+
+    public VacuumDiscovery() throws IllegalArgumentException {
+        super(15);
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Start Xiaomi Robot Vacuum background discovery");
+        if (roboDiscoveryJob == null || roboDiscoveryJob.isCancelled()) {
+            roboDiscoveryJob = scheduler.scheduleWithFixedDelay(() -> discover(), 0, SEARCH_INTERVAL, TimeUnit.SECONDS);
+        }
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        logger.debug("Stop Xiaomi Robot Vacuum background discovery");
+        if (roboDiscoveryJob != null && !roboDiscoveryJob.isCancelled()) {
+            roboDiscoveryJob.cancel(true);
+            roboDiscoveryJob = null;
+        }
+    }
+
+    private synchronized void discover() {
+        TreeSet<String> broadcastAdresses = getBroadcastAddresses();
+        HashMap<String, byte[]> responses = new HashMap<String, byte[]>();
+        for (String broadcastAdress : broadcastAdresses) {
+            responses.putAll(sendDiscoveryRequest(broadcastAdress));
+        }
+        for (Entry<String, byte[]> i : responses.entrySet()) {
+            logger.trace("Discovery responses from : {}:{}", i.getKey(), Utils.getSpacedHex(i.getValue()));
+            Message msg = new Message(i.getValue());
+            logger.debug("Vacuum time stamp: {}, OH time {}, delta {}", msg.getTimestamp(), LocalDateTime.now(),
+                    LocalDateTime.now().compareTo(msg.getTimestamp()));
+            String token = Utils.getHex(msg.getChecksum());
+            String id = Utils.getHex(msg.getDeviceId());
+
+            ThingUID uid = new ThingUID(XiaomiVacuumBindingConstants.THING_TYPE_MIIO, id);
+            // TODO: placeholder entries. Here test for MiIO things to determine as what thingType it needs to be added
+            if (discoveryServiceCallback.getExistingDiscoveryResult(uid) != null) {
+                logger.debug("Thing {} was already in discovery Inbox", uid.toString());
+            }
+            if (discoveryServiceCallback.getExistingThing(uid) != null) {
+                logger.debug("Thing {} already exists", uid.toString());
+            }
+            logger.debug("Discovered Xiaomi Robot Vacuum {} at {}", id, i.getKey());
+            if (token.equals("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF") || token.equals("00000000000000000000000000000000")) {
```

Comment:
Creating a List<String> of the invalid tokens and using contains on it would make this code a bit more readable.